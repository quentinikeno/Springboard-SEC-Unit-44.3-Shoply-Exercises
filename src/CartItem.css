.CartItem i {
	cursor: pointer;
}

.CartItem .image img {
	width: fit-content;
	height: fit-content;
	max-width: 128px;
	max-height: 128px;
}
