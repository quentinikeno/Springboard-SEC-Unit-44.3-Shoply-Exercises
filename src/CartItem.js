import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import useAddRemoveCart from "./hooks/useAddRemoveCart";

const CartItem = ({ id, quantity }) => {
	const product = useSelector(
		(state) => state.products.inventory.products[id]
	);

	const [handleAddToCart, handleRemoveFromCart] = useAddRemoveCart(id);

	const { name, price, image_url } = product;

	return (
		<div className="box level has-text-centered">
			<figure className="level-item">
				<img className="image is-128x128" src={image_url} alt={name} />
			</figure>

			<div className="level-item is-capitalized">
				<div>
					<p className="heading">Product</p>
					<p className="title">
						<Link to={`/products/${id}`}>{name}</Link>
					</p>
				</div>
			</div>
			<div className="level-item">
				<div>
					<p className="heading">Price</p>
					<p className="subtitle">${price}</p>
				</div>
			</div>
			<div className="level-item">
				<div>
					<p className="heading">Qty</p>

					<i
						className="fa-solid fa-angle-up"
						onClick={handleAddToCart}
					></i>
					<span className="mx-3 subtitle">{quantity}</span>
					<i
						className="fa-solid fa-chevron-down"
						onClick={handleRemoveFromCart}
					></i>
				</div>
			</div>
		</div>
	);
};

export default CartItem;
