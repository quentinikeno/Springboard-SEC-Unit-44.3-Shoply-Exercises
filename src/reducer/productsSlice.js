import { createSlice } from "@reduxjs/toolkit";
import inventory from "../data.json";

const initialState = {
	inventory,
	cart: {},
};

export const productsSlice = createSlice({
	name: "products",
	initialState,
	reducers: {
		addToCart: (state, action) => {
			const { id, qty } = action.payload;
			const foundProduct = state.cart[id];
			if (foundProduct) {
				state["cart"][id]["quantity"] += qty;
			} else {
				state["cart"][id] = { quantity: qty };
			}
		},
		removeFromCart: (state, action) => {
			const { id, qty } = action.payload;
			const foundProduct = state.cart[id];
			if (foundProduct) {
				state["cart"][id]["quantity"] -= qty;
				if (state["cart"][id]["quantity"] < 1) delete state["cart"][id];
			} else {
				console.error("Item not found in cart.");
			}
		},
		clearCart: (state) => {
			state.cart = {};
		},
	},
});

export const { addToCart, removeFromCart, clearCart } = productsSlice.actions;

export default productsSlice.reducer;
