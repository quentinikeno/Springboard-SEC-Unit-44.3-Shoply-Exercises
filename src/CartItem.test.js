import { render } from "@testing-library/react";
import { store } from "./reducer/store";
import { Provider } from "react-redux";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import CartItem from "./CartItem";

const tvId = "2266b2d6-3d71-4356-8065-392f1fceab07";

it("renders without crashing", () => {
	render(
		<Provider store={store}>
			<MemoryRouter initialEntries={["/cart"]}>
				<Routes>
					<Route
						path="/cart"
						element={<CartItem id={tvId} quantity={1} />}
					></Route>
				</Routes>
			</MemoryRouter>
		</Provider>
	);
});

it("renders matches snapshot", () => {
	const { asFragment } = render(
		<Provider store={store}>
			<MemoryRouter initialEntries={["/cart"]}>
				<Routes>
					<Route
						path="/cart"
						element={<CartItem id={tvId} quantity={1} />}
					></Route>
				</Routes>
			</MemoryRouter>
		</Provider>
	);
	expect(asFragment()).toMatchSnapshot();
});
