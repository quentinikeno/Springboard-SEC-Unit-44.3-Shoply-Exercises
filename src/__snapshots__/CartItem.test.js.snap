// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CartItem box level has-text-centered"
  >
    <figure
      class="level-item image is-128x128"
    >
      <img
        alt="phone"
        src="https://images-na.ssl-images-amazon.com/images/I/717DO5Q-aCL._SY879_.jpg"
      />
    </figure>
    <div
      class="level-item is-capitalized"
    >
      <div>
        <p
          class="heading"
        >
          Product
        </p>
        <p
          class="title"
        >
          <a
            href="/products/2266b2d6-3d71-4356-8065-392f1fceab07"
          >
            phone
          </a>
        </p>
      </div>
    </div>
    <div
      class="level-item"
    >
      <div>
        <p
          class="heading"
        >
          Price
        </p>
        <p
          class="subtitle"
        >
          $399.99
        </p>
      </div>
    </div>
    <div
      class="level-item"
    >
      <div>
        <p
          class="heading"
        >
          Qty
        </p>
        <i
          class="fa-solid fa-angle-up"
        />
        <span
          class="mx-3 subtitle"
        >
          1
        </span>
        <i
          class="fa-solid fa-chevron-down"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
